core/common/src/main/java/alluxio/network/connection/ThriftClientPool.java

```
@@ -0,0 +1,216 @@
+   *
+   * @return the thrift client created
+   * @throws IOException if it fails to create a thrift client
+   */
+  @Override
+  protected T createNewResource() throws IOException {
+    TTransport transport = mTransportProvider.getClientTransport(mAddress);
+    TProtocol binaryProtocol = new TBinaryProtocol(transport);
+    T client = createThriftClient(new TMultiplexedProtocol(binaryProtocol, mServiceName));
+
+    RetryPolicy retry =
+        new ExponentialBackoffRetry(CONNECTION_OPEN_RETRY_BASE_SLEEP_MS, Constants.SECOND_MS,
+            CONNECTION_OPEN_RETRY_MAX);
+    while (true) {
+      try {
+        if (!transport.isOpen()) {
+          transport.open();
+        }
+        if (transport.isOpen()) {
+          checkVersion(client);
+        }
+      } catch (TTransportException e) {
+        LOG.error(
+            "Failed to connect (" + retry.getRetryCount() + ") to " + getServiceNameForLogging()
+                + " @ " + mAddress, e);
+        if (!retry.attemptRetry()) {
+          throw new IOException(e);
+        }
+      }
+      break;
+    }
+
+    return client;
+  }
+
+  /**
+   * Checks whether a client is healthy.
+   *
+   * @param client the thrift client to check
+   * @return true if the client is open (i.e. connected)
+   */
+  @Override
+  protected boolean isHealthy(T client) {
+    return client.getOutputProtocol().getTransport().isOpen();
+  }
+
+  @Override
+  protected boolean shouldGc(ResourceInternal<T> clientResourceInternal) {
+    return System.currentTimeMillis() - clientResourceInternal.getLastAccessTimeMs()
+        > (long) mGcThresholdInSecs * (long) Constants.SECOND_MS;
+  }
+
+  /**
+   * Check the service version to see whether it matches the expected version.
+   *
+   * @param client the client
+   * @throws IOException if it fails to check version
+   */
+  private void checkVersion(T client) throws TTransportException, IOException {
+    if (mIsVersionChecked) {
+      return;
+    }
```

Done. I've done what you suggested. I realize that it won't fail fast after I always set this mIsVersionChecked when it fails. So what you suggested is cleaner.